File: composeApp/src/commonMain/kotlin/io/awairo/presentation/theme/SkyTheme.kt
package io.awairo.presentation.theme

import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.graphics.Color
import io.awairo.domain.model.SkyPalette
import io.awairo.domain.model.ThemeMode

data class SkyTheme(
    val palette: SkyPalette,
    val mode: ThemeMode,
    val isDark: Boolean,
    val backgroundTop: Color,
    val backgroundBottom: Color,
    val gradientA: Color,
    val gradientB: Color,
    val gradientC: Color,
    val textPrimary: Color,
    val textSecondary: Color,
)

val LocalSkyTheme = compositionLocalOf<SkyTheme> {
    error("SkyTheme not provided. Wrap your composable in CompositionLocalProvider(LocalSkyTheme provides ...)")
}

/**
 * 現在のパレット・ダークモードフラグから [SkyTheme] を解決する。
 */
fun resolveSkyTheme(palette: SkyPalette, mode: ThemeMode, isSystemDark: Boolean): SkyTheme {
    val dark = when (mode) {
        ThemeMode.SYSTEM -> isSystemDark
        ThemeMode.DARK -> true
        ThemeMode.LIGHT -> false
    }
    val colors = if (dark) darkColorsFor(palette) else lightColorsFor(palette)
    return SkyTheme(
        palette = palette,
        mode = mode,
        isDark = dark,
        backgroundTop = colors.bgTop,
        backgroundBottom = colors.bgBottom,
        gradientA = colors.gradA,
        gradientB = colors.gradB,
        gradientC = colors.gradC,
        textPrimary = colors.textPrimary,
        textSecondary = colors.textSecondary,
    )
}

private data class PaletteColors(
    val bgTop: Color, val bgBottom: Color,
    val gradA: Color, val gradB: Color, val gradC: Color,
    val textPrimary: Color, val textSecondary: Color,
)

private fun darkColorsFor(p: SkyPalette): PaletteColors = when (p) {
    SkyPalette.NIGHT_SKY -> PaletteColors(
        Color(0xFF03030E), Color(0xFF06061A),
        Color(0xD9190837), Color(0xCC081437), Color(0x8C041E16),
        Color(0xCCFFFFFF), Color(0x80FFFFFF)
    )
    SkyPalette.MIST -> PaletteColors(
        Color(0xFF070C12), Color(0xFF0A1220),
        Color(0xD908193C), Color(0xCC052832), Color(0x8C031E2D),
        Color(0xCCFFFFFF), Color(0x80FFFFFF)
    )
    SkyPalette.DUSK -> PaletteColors(
        Color(0xFF0E0608), Color(0xFF1C0A0A),
        Color(0xD9370C08), Color(0xCC280F05), Color(0x8C0A061E),
        Color(0xCCFFFFFF), Color(0x80FFFFFF)
    )
    SkyPalette.KOMOREBI -> PaletteColors(
        Color(0xFF040C06), Color(0xFF061410),
        Color(0xD9062312), Color(0xCC08321E), Color(0x801E2808),
        Color(0xCCFFFFFF), Color(0x80FFFFFF)
    )
    SkyPalette.AKATSUKI -> PaletteColors(
        Color(0xFF0A0608), Color(0xFF180A14),
        Color(0xD9320828), Color(0xCC0C0837), Color(0x8C18081E),
        Color(0xCCFFFFFF), Color(0x80FFFFFF)
    )
    SkyPalette.SILVER_FOG -> PaletteColors(
        Color(0xFF080A10), Color(0xFF10121E),
        Color(0xD9141932), Color(0xCC1E233C), Color(0x8C12172A),
        Color(0xCCFFFFFF), Color(0x80FFFFFF)
    )
}

private fun lightColorsFor(p: SkyPalette): PaletteColors = when (p) {
    SkyPalette.NIGHT_SKY -> PaletteColors(
        Color(0xFFECEAF4), Color(0xFFDDDAF0),
        Color(0x40A08CDC), Color(0x338CA0E6), Color(0x00000000),
        Color(0xCC000000), Color(0x80000000)
    )
    SkyPalette.MIST -> PaletteColors(
        Color(0xFFEAF0F4), Color(0xFFD8E8F0),
        Color(0x478CB4D2), Color(0x38A0C8DC), Color(0x00000000),
        Color(0xCC000000), Color(0x80000000)
    )
    SkyPalette.DUSK -> PaletteColors(
        Color(0xFFF8EDE4), Color(0xFFF0DDD0),
        Color(0x52F0A064), Color(0x38DC7850), Color(0x29B464A0),
        Color(0xCC000000), Color(0x80000000)
    )
    SkyPalette.KOMOREBI -> PaletteColors(
        Color(0xFFEEF4E8), Color(0xFFE0ECDA),
        Color(0x478CC864), Color(0x38B4DC78), Color(0x2EDCC850),
        Color(0xCC000000), Color(0x80000000)
    )
    SkyPalette.AKATSUKI -> PaletteColors(
        Color(0xFFF8EAF0), Color(0xFFF0D8E8),
        Color(0x47DC8CB4), Color(0x33B48CDC), Color(0x00000000),
        Color(0xCC000000), Color(0x80000000)
    )
    SkyPalette.SILVER_FOG -> PaletteColors(
        Color(0xFFEEEFF4), Color(0xFFE0E2EC),
        Color(0x4DB4B9D2), Color(0x38A0A5C8), Color(0x00000000),
        Color(0xCC000000), Color(0x80000000)
    )
}

File: packages/AwaIroPresentation/Sources/AwaIroPresentation/Theme/SkyTheme.swift
import AwaIroDomain
import SwiftUI

public struct SkyTheme: Equatable, Sendable {
  public let palette: SkyPalette
  public let mode: ThemeMode
  public let isDark: Bool
  public let backgroundTop: Color
  public let backgroundBottom: Color
  public let textPrimary: Color
  public let textSecondary: Color

  public init(palette: SkyPalette, mode: ThemeMode, systemColorScheme: ColorScheme) {
    self.palette = palette
    self.mode = mode

    let resolvedDark: Bool
    switch mode {
    case .system: resolvedDark = systemColorScheme == .dark
    case .dark: resolvedDark = true
    case .light: resolvedDark = false
    }
    self.isDark = resolvedDark

    let colors = Self.resolve(palette: palette, isDark: resolvedDark)
    self.backgroundTop = colors.top
    self.backgroundBottom = colors.bottom
    self.textPrimary = resolvedDark ? .white : .black
    self.textSecondary = resolvedDark ? Color.white.opacity(0.6) : Color.black.opacity(0.55)
  }

  /// Default theme used when env not provided (e.g. previews / snapshots).
  public static let `default` = SkyTheme(
    palette: .nightSky, mode: .dark, systemColorScheme: .dark)

  private static func resolve(palette: SkyPalette, isDark: Bool) -> (top: Color, bottom: Color) {
    switch (palette, isDark) {
    case (.nightSky, true):
      return (Color(red: 0.06, green: 0.05, blue: 0.18), Color(red: 0.10, green: 0.10, blue: 0.30))
    case (.nightSky, false):
      return (Color(red: 0.78, green: 0.74, blue: 0.92), Color(red: 0.92, green: 0.90, blue: 0.98))
    case (.mist, true):
      return (Color(red: 0.05, green: 0.10, blue: 0.18), Color(red: 0.08, green: 0.16, blue: 0.26))
    case (.mist, false):
      return (Color(red: 0.78, green: 0.84, blue: 0.92), Color(red: 0.90, green: 0.94, blue: 0.98))
    case (.dusk, true):
      return (Color(red: 0.18, green: 0.06, blue: 0.06), Color(red: 0.30, green: 0.10, blue: 0.10))
    case (.dusk, false):
      return (Color(red: 1.00, green: 0.78, blue: 0.74), Color(red: 1.00, green: 0.88, blue: 0.82))
    case (.komorebi, true):
      return (Color(red: 0.05, green: 0.18, blue: 0.06), Color(red: 0.08, green: 0.26, blue: 0.10))
    case (.komorebi, false):
      return (Color(red: 0.78, green: 0.92, blue: 0.74), Color(red: 0.90, green: 0.98, blue: 0.84))
    case (.akatsuki, true):
      return (Color(red: 0.18, green: 0.05, blue: 0.18), Color(red: 0.30, green: 0.08, blue: 0.30))
    case (.akatsuki, false):
      return (Color(red: 1.00, green: 0.84, blue: 0.92), Color(red: 1.00, green: 0.92, blue: 0.96))
    case (.silverFog, true):
      return (Color(red: 0.10, green: 0.10, blue: 0.18), Color(red: 0.14, green: 0.14, blue: 0.22))
    case (.silverFog, false):
      return (Color(red: 0.85, green: 0.85, blue: 0.88), Color(red: 0.94, green: 0.94, blue: 0.96))
    }
  }
}

private struct SkyThemeKey: EnvironmentKey {
  static let defaultValue: SkyTheme = .default
}

extension EnvironmentValues {
  public var skyTheme: SkyTheme {
    get { self[SkyThemeKey.self] }
    set { self[SkyThemeKey.self] = newValue }
  }
}

extension View {
  public func skyTheme(_ theme: SkyTheme) -> some View {
    self.environment(\.skyTheme, theme)
  }
}
